description: listIndexes-serverErrors

schemaVersion: '1.9'

runOnRequirements:
  -
    minServerVersion: '4.0'
    topologies:
      - single
      - replicaset
  -
    minServerVersion: 4.1.7
    topologies:
      - sharded
      - load-balanced

createEntities:
  -
    client:
      id: &client0 client0
      useMultipleMongoses: false
      observeEvents:
        - commandStartedEvent
  -
    database:
      id: &database0 database0
      client: *client0
      databaseName: &database_name retryable-reads-tests
  -
    collection:
      id: &collection0 collection0
      database: *database0
      collectionName: &collection_name coll

initialData:
  -
    collectionName: *collection_name
    databaseName: *database_name
    documents: []

tests:
  -
    description: 'ListIndexes succeeds after InterruptedAtShutdown'
    operations:
      -
        object: testRunner
        name: failPoint
        arguments:
          client: *client0
          failPoint: &failCommand_failPoint
            configureFailPoint: failCommand
            mode:
              times: 1
            data:
              failCommands:
                - listIndexes
              errorCode: 11600
      - &retryable_operation
        object: *collection0
        name: listIndexes
    expectEvents:
      -
        client: *client0
        events:
          - &retryable_command_started_event
            commandStartedEvent:
              command:
                listIndexes: *collection_name
              databaseName: *database_name
          - *retryable_command_started_event
  -
    description: 'ListIndexes succeeds after InterruptedDueToReplStateChange'
    operations:
      -
        object: testRunner
        name: failPoint
        arguments:
          client: *client0
          failPoint:
            configureFailPoint: failCommand
            mode:
              times: 1
            data:
              failCommands:
                - listIndexes
              errorCode: 11602
      - *retryable_operation
    expectEvents:
      -
        client: *client0
        events:
          - *retryable_command_started_event
          - *retryable_command_started_event
  -
    description: 'ListIndexes succeeds after NotWritablePrimary'
    operations:
      -
        object: testRunner
        name: failPoint
        arguments:
          client: *client0
          failPoint:
            configureFailPoint: failCommand
            mode:
              times: 1
            data:
              failCommands:
                - listIndexes
              errorCode: 10107
      - *retryable_operation
    expectEvents:
      -
        client: *client0
        events:
          - *retryable_command_started_event
          - *retryable_command_started_event
  -
    description: 'ListIndexes succeeds after NotPrimaryNoSecondaryOk'
    operations:
      -
        object: testRunner
        name: failPoint
        arguments:
          client: *client0
          failPoint:
            configureFailPoint: failCommand
            mode:
              times: 1
            data:
              failCommands:
                - listIndexes
              errorCode: 13435
      - *retryable_operation
    expectEvents:
      -
        client: *client0
        events:
          - *retryable_command_started_event
          - *retryable_command_started_event
  -
    description: 'ListIndexes succeeds after NotPrimaryOrSecondary'
    operations:
      -
        object: testRunner
        name: failPoint
        arguments:
          client: *client0
          failPoint:
            configureFailPoint: failCommand
            mode:
              times: 1
            data:
              failCommands:
                - listIndexes
              errorCode: 13436
      - *retryable_operation
    expectEvents:
      -
        client: *client0
        events:
          - *retryable_command_started_event
          - *retryable_command_started_event
  -
    description: 'ListIndexes succeeds after PrimarySteppedDown'
    operations:
      -
        object: testRunner
        name: failPoint
        arguments:
          client: *client0
          failPoint:
            configureFailPoint: failCommand
            mode:
              times: 1
            data:
              failCommands:
                - listIndexes
              errorCode: 189
      - *retryable_operation
    expectEvents:
      -
        client: *client0
        events:
          - *retryable_command_started_event
          - *retryable_command_started_event
  -
    description: 'ListIndexes succeeds after ShutdownInProgress'
    operations:
      -
        object: testRunner
        name: failPoint
        arguments:
          client: *client0
          failPoint:
            configureFailPoint: failCommand
            mode:
              times: 1
            data:
              failCommands:
                - listIndexes
              errorCode: 91
      - *retryable_operation
    expectEvents:
      -
        client: *client0
        events:
          - *retryable_command_started_event
          - *retryable_command_started_event
  -
    description: 'ListIndexes succeeds after HostNotFound'
    operations:
      -
        object: testRunner
        name: failPoint
        arguments:
          client: *client0
          failPoint:
            configureFailPoint: failCommand
            mode:
              times: 1
            data:
              failCommands:
                - listIndexes
              errorCode: 7
      - *retryable_operation
    expectEvents:
      -
        client: *client0
        events:
          - *retryable_command_started_event
          - *retryable_command_started_event
  -
    description: 'ListIndexes succeeds after HostUnreachable'
    operations:
      -
        object: testRunner
        name: failPoint
        arguments:
          client: *client0
          failPoint:
            configureFailPoint: failCommand
            mode:
              times: 1
            data:
              failCommands:
                - listIndexes
              errorCode: 6
      - *retryable_operation
    expectEvents:
      -
        client: *client0
        events:
          - *retryable_command_started_event
          - *retryable_command_started_event
  -
    description: 'ListIndexes succeeds after NetworkTimeout'
    operations:
      -
        object: testRunner
        name: failPoint
        arguments:
          client: *client0
          failPoint:
            configureFailPoint: failCommand
            mode:
              times: 1
            data:
              failCommands:
                - listIndexes
              errorCode: 89
      - *retryable_operation
    expectEvents:
      -
        client: *client0
        events:
          - *retryable_command_started_event
          - *retryable_command_started_event
  -
    description: 'ListIndexes succeeds after SocketException'
    operations:
      -
        object: testRunner
        name: failPoint
        arguments:
          client: *client0
          failPoint:
            configureFailPoint: failCommand
            mode:
              times: 1
            data:
              failCommands:
                - listIndexes
              errorCode: 9001
      - *retryable_operation
    expectEvents:
      -
        client: *client0
        events:
          - *retryable_command_started_event
          - *retryable_command_started_event
  -
    description: 'ListIndexes fails after two NotWritablePrimary errors'
    operations:
      -
        object: testRunner
        name: failPoint
        arguments:
          client: *client0
          failPoint:
            configureFailPoint: failCommand
            mode:
              times: 2
            data:
              failCommands:
                - listIndexes
              errorCode: 10107
      - &retryable_operation_fails
        object: *collection0
        name: listIndexes
        expectError:
          isError: true
    expectEvents:
      -
        client: *client0
        events:
          - *retryable_command_started_event
          - *retryable_command_started_event
  -
    description: 'ListIndexes fails after NotWritablePrimary when retryReads is false'
    operations:
      -
        object: testRunner
        name: createEntities
        arguments:
          entities:
            - client:
                id: &client1 client1
                useMultipleMongoses: false
                uriOptions:
                  retryReads: false
                observeEvents:
                  - commandStartedEvent
            - database:
                id: &database1 database1
                client: *client1
                databaseName: *database_name
            - collection:
                id: &collection1 collection1
                database: *database1
                collectionName: *collection_name
      -
        object: testRunner
        name: failPoint
        arguments:
          client: *client1
          failPoint:
            configureFailPoint: failCommand
            mode:
              times: 1
            data:
              failCommands:
                - listIndexes
              errorCode: 10107
      - <<: *retryable_operation_fails
        object: *collection1
    expectEvents:
      -
        client: *client1
        events:
          - *retryable_command_started_event
